crates/sui-core/src/consensus_adapter.rs

```diff
@@ -284,6 +353,31 @@ impl ConsensusAdapter {
         processed_waiter
             .await
             .expect("Storage error when waiting for consensus message processed");
+        let send_end_of_publish =
+            if let ConsensusTransactionKind::UserTransaction(certificate) = &transaction.kind {
+                let reconfig_guard = self.authority.get_reconfig_state_read_lock_guard();
+                // note that pending_certificates lock is always acquired *after* reconfiguration lock
+                // acquiring locks in different order might lead to deadlocks
+                let mut pending_certificates = self.pending_certificates.lock();
+                pending_certificates.remove(certificate.digest().as_ref());
```

Comment:
So this removal happens out of sync with the removal from pending_consensus_transactions?

Reply:
Yes, which I think is ok - in memory `pending_certificates` is only source of truth when it comes to making reconfiguration decisions. And the persistent table is updated 'some time after'

Reply:
I wonder if we shouldn't remove the transaction from the pending_consensus_transactions table if the validator has closed the epoch (i.e. reject all certs). Because pending_consensus_transactions would be the table where we find all certs that need to be reverted in the ened?

Reply:
Thinking about this, perhaps it should work as following:

* The transition to RejectAllCerts happens inside consensus handler. Once this has happened, we do not process any user certificates from consensus, so this code path we are commenting about is never activated (so no need for this check here).
* Once we start working on revert, we'll need some mechanism to kill those consensus waiting tasks to clean up memory.

What do you think?

Reply:
Right now we already have a check in `handle_consensus_transaction`, is that sufficient?

Reply:
Yes!